core/src/streams.rs

```diff
@@ -285,10 +327,434 @@ impl<'gc> NetStream<'gc> {
+                        (_, FlvSoundRate::R11_000) => 11_000,
+                        (_, FlvSoundRate::R22_000) => 22_000,
+                        (_, FlvSoundRate::R44_000) => 44_000,
+                    },
+                    is_stereo: match audio_data.sound_type {
+                        FlvSoundType::Mono => false,
+                        FlvSoundType::Stereo => true,
+                    },
+                    is_16_bit: match audio_data.size {
+                        FlvSoundSize::Bits8 => false,
+                        FlvSoundSize::Bits16 => true,
+                    },
+                };
+
+                let sound_stream_head = SoundStreamInfo {
+                    wrapping: SoundStreamWrapping::Unwrapped,
+                    stream_format: swf_format,
+                    num_samples_per_block: 0,
+                    latency_seek: 0,
+                };
+
+                *audio_stream = Some((substream, sound_stream_head));
+
+                &mut audio_stream.as_mut().unwrap().0
+            }
+        };
+
+        Ok(substream.append(data)?)
+    }
+
+    /// Determine if the given sound is currently playing.
+    fn sound_currently_playing(
+        context: &mut UpdateContext<'_, 'gc>,
+        sound: &Option<SoundInstanceHandle>,
+    ) -> bool {
+        sound
+            .map(|si| context.audio.is_sound_playing(si))
+            .unwrap_or(false)
+    }
+
+    /// Clean up after a sound instance that has finished playing.
+    ///
+    /// Generally speaking, streams are only to be used once. However, the
+    /// audio backend will only retain information about sounds that are
+    /// currently playing, so if the sound has finished since the last tick, we
+    /// need to restart it.
+    ///
+    /// Intended to be called at the start of tag processing, before any new
+    /// audio data has been streamed.
+    fn collect_sound_stream(
+        self,
+        context: &mut UpdateContext<'_, 'gc>,
+        write: &mut NetStreamData<'gc>,
+    ) {
+        if !Self::sound_currently_playing(context, &write.sound_instance) {
+            write.audio_stream = None;
+            write.sound_instance = None;
+        }
+    }
+
+    /// Ensure that if we have queued up audio into a sound stream, that said
+    /// stream gets sent over to the audio backend.
+    ///
+    /// Intended to be called at the end of tag processing. Audio processing
+    /// should occur only after a minimum number of tags have been processed to
+    /// avoid audio underruns.
+    fn commit_sound_stream(
+        self,
+        context: &mut UpdateContext<'_, 'gc>,
+        write: &mut NetStreamData<'gc>,
+    ) -> Result<(), NetstreamError> {
+        if !Self::sound_currently_playing(context, &write.sound_instance) {
+            if let Some((substream, sound_stream_head)) = &mut write.audio_stream {
+                let attached_to = write.attached_to;
+
+                if context.is_action_script_3() {
```

Comment:
I'd prefer this to check if `avm_object` is `Some(Avm2Object(...))` instead of using `context.is_action_script_3`, for mixed AVM.

Reply:
(Though I have no knowledge of the audio manager- if this does actually depend on the root movie type, `context.swf.is_action_script_3` is fine).

Reply:
I thought this was the correct way to check - and we check `context.is_action_script_3()` in at least 90 places. I specifically remember us removing `avm_object` checks for context type checks; if this is going to be a pain for mixed AVM we probably should be having a separate PR to remove these checks separately.